### ISSUE TITLE
updateSelectInput() not working on UI rendered in modal dialog box

### ISSUE THREAD

**Issue by mmccarthy404** [rstudio/shiny#2744]
I have a module with both a UI component and a server component. The UI contains a selectInput widget that is updated within the server component using updateSelectInput(). I've used this logic in teh past without issue, however, this time I'm attempting to render the UI component within a modal dialog box, and can't get the selectInput widget to update.

As a small example of what I'm talking about, I've created the following app with a module, select.R, with a UI component select_module_ui() called both inside and outside of a modal dialog box, however, only the UI component rendered outside of a modal dialog box is updated with updateSelectInput().

**ui.R**
```
# ui.R

fluidPage(
  actionButton("popup",
               label = "Pop Up"),
  tags$hr(),
  select_module_ui("select")
)
```

**server.R**
```
# server.R

function(input, output, session) {
  
  # Handling select backend logic.
  callModule(select_module, "select")
  
  # Rendering modal UI.
  observeEvent(input$popup, {
    showModal(
      modalDialog(
        select_module_ui("select")
      )
    )
  })
  
}
```

**global.R**
```
# global.R

source("modules/select.R")
```

**modules/select.R**
```
# modules/select.R

select_module_ui <- function(id) {
  
  ns <- NS(id)
  
  tagList(
    selectInput(ns("input_select"),
                label = "Select Number:",
                choices = c(1, 2, 3))
  )

}

select_module <- function(input, output, session) {

  # Updating choices of "input_select".
  updateSelectInput(session,
                    "input_select",
                    choices = c(4, 5, 6))
  
}
```

**Comment by cpsievert** [comment]
> can't get the selectInput widget to update.

Can you be more specific about what exactly you trying to achieve? That is, what user events should lead to which selectInput() having which values?

**Comment by hedjour** [comment]
Hi I  have the same bug with this ECM code :
```r
mod_graph1var_ui <- function(id){
  ns <- NS(id) #Redéfinit la fonction ns
  tagList(
    pageWithSidebar(
      # title
      headerPanel("Select Options"),
      #input
      sidebarPanel(
        selectInput(ns("dataset"),  "Dataset" , "NONE", selected = "NONE"),  
        selectInput(ns("variable"), "Variable", "NONE", selected = "NONE"),
        selectInput(ns("group"),    "Group"   , "NONE", selected = "NONE")
      ),
      # output                
      mainPanel(
        h3(textOutput("caption")),
        uiOutput("plot") # depends on input 
      )
    )
  )
}

#' graph1var Server Functions
mod_graph1var_server <- function(id, my_tables){
  moduleServer(id, function(input, output, session){
    ns <- session$ns
    shiny::observe({
      print("On change les menus^")
      #update variable and group based on dataset
      output$dataset <- updateSelectInput(session, "dataset_in",
                                          # label = "Vue:",
                                          choices = names(my_tables) )})
    
    shiny::observe({
      #update variable and group based on dataset
      if(!is.null(input$dataset_in)){
      output$variable <- updateSelectInput(session, "variable",
                            # label = "Variable:",
                            choices = colnames(my_tables[[input$dataset_in]]) )
      output$group <- updateSelectInput(session, "group",
                            # label = "Groups:",
                            choices = colnames( my_tables[[input$dataset_in]]) )

      output$caption<- shiny::renderText({
        switch(input$plot_type,
               "boxplot" 	 = 	"Boxplot",
               "histogram" =	"Histogram",
               "density" 	 =	"Density plot",
               "bar" 		   =	"Bar graph") })
      }#Fin du if
    })#Fin du observe
    
  }) #Close module server
}# Close definition of global function : mod_graph1var_server

app_ui <- function(request) {
  # Your application UI logic 
  shinydashboard::dashboardPage(
    shinydashboard::dashboardHeader(title = "Analysis BDD"),
    shinydashboard::dashboardSidebar(
      shinydashboard::sidebarMenu(
        menuItem("1var Graphiques", tabName = "graph1var", icon = icon("poll"))
      )
    ),
    shinydashboard::dashboardBody(
      tabItems(
        # 1var graph
        tabItem(tabName = "graph1var",
                h1("Graphiques with 1 variables"),
                mod_graph1var_ui("graph1var_ui_1")
        )
      )
    )
  )
}
app_server <- function( input, output, session ) {
  # Your application server logic 
  mytables <- list(test1 = data.frame(id = seq(1,100,1), mes = runif(n = 100,0,1)), 
                   test2 = data.frame(id = seq(1,100,1), mes = runif(n = 100,0,1)) 
  )
  mod_graph1var_server("graph1var_ui_1", my_tables = mytables)
}

library(shiny)
library(shinydashboard)
shinyApp(ui=app_ui, server = app_server)
```
I wish to fill the select input from names of tables from a DB in my real app. The DB is represented by mytables var in this ECM

**Comment by Excidion** [comment]
I am running into the same problem where a `selectizeInput`, that is part of a `modalDialog` can not be updated via `updateSelectizeInput` as described [here](https://shiny.rstudio.com/articles/selectize.html).

**Comment by codyevers** [comment]
@Excidion, I was able to update the select input from within the modalDialog function, so I believe this issue is related to variable scoping.

**Comment by iamthelaw-636** [comment]
Can you elaborate on that @codyevers? Did anyone get any closer to solving this riddle? I think I've run into the same problem.

**Comment by alon-sarid** [comment]
I bypassed this issue by not using updateSelectizeInput  with modal.
in my case the modal was  called in observe. prior to the modal call, 
I updated all the dynamic fields. of course this is not a real solution, only a limited bypass

**Comment by ShixiangWang** [comment]
Try calling the `callModule(select_module, "select")` in the `showModal` domain

**Comment by gadenbuie** [comment]
Because this old issue was bumped and because it seems to have affected quite a few people, I went back and took a look at the original issue.

The goal, as far as I understand it, is to update a select input (or any input really) that's dynamically created inside a modal dialog. Modal dialogues are interesting because the inputs don't exist until the modal appears on the page. The dynamic UI elements are also re-created each time the modal is shown again.

First a quick problem in the original reprex: `callModule(select_module, "select")` is called once, but there are two calls to `select_module_ui("select")`. What's needed instead are paired ui/module calls, one to create a `"select_ui"` instance and one to create a `"select_modal"` instance.

That said, the core problem I suspect many people in this thread are running into is that you cannot call `updateSelectInput()` to update the element in the modal, until you know that the modal has been shown. So @ShixiangWang's answer is generally right: call `updateSelectInput()` after calling `showModal()`. You could call `callModule(select_module, "select_modal")` just after `showModal()`, or you can make the `updateSelectInput()` reactive.

Here's how you'd do the second approach. In the `select_module()` server function, I'll add a `trigger_update` parameter. If it's `NULl`, we'll initialize it ourselves, otherwise we could let the user pass in a reactive trigger. We'll also return the `trigger_update` object from the server function, which gives us a way to trigger the update in our parent server logic.

```r
select_module <- function(input, output, session, trigger_update = NULL) {
  if (is.null(trigger_update)) {
    trigger_update <- reactiveVal(0)
  }

  observeEvent(trigger_update(), {
    # Updating choices of "input_select".
    updateSelectInput(
      session,
      "input_select",
      choices = c(4, 5, 6)
    )
  })

  trigger_update
}
```

Then, in the global server function, I'll initialize the two select modules, but I'll save the return value for the one shown in the modal:

```r
server <- function(input, output, session) {

  # Handling select backend logic.
  callModule(select_module, "select_ui")

  # The module returns a reactive whose value triggers an update of the select
  update_modal_select <- callModule(select_module, "select_modal")

  # <!-- Modal UI rendering logic will be here -->

}
```

Finally, when we create the modal, we can also trigger an update:

```r
  # Rendering modal UI.
  observeEvent(input$popup, {
    showModal(
      modalDialog(
        select_module_ui("select_modal")
      )
    )
    update_modal_select(update_modal_select() + 1)
  })
```

Here's the whole working reprex app in a single R script:

<details>
<summary>`app.R`</summary>

```r
library(shiny)

# modules/select.R ---------------------------------------
select_module_ui <- function(id) {
  ns <- NS(id)

  tagList(
    selectInput(
      ns("input_select"),
      label = "Select Number:",
      choices = c(1, 2, 3)
    )
  )
}

select_module <- function(input, output, session, trigger_update = NULL) {
  if (is.null(trigger_update)) {
    trigger_update <- reactiveVal(0)
  }

  observeEvent(trigger_update(), {
    # Updating choices of "input_select".
    updateSelectInput(
      session,
      "input_select",
      choices = c(4, 5, 6)
    )
  })

  trigger_update
}

# UI ---------------------------------------
ui <- fluidPage(
  actionButton("popup", label = "Pop Up"),
  tags$hr(),
  select_module_ui("select_ui")
)

# Server ---------------------------------------
server <- function(input, output, session) {

  # Handling select backend logic.
  callModule(select_module, "select_ui")

  # The module returns a reactive whose value triggers an update of the select
  update_modal_select <- callModule(select_module, "select_modal")

  # Rendering modal UI.
  observeEvent(input$popup, {
    showModal(
      modalDialog(
        select_module_ui("select_modal")
      )
    )
    update_modal_select(update_modal_select() + 1)
  })

}

shinyApp(ui, server)
```

</details>

**Comment by AlexIvanHoward** [comment]
I stumbled upon this post last week when I encountered a similar problem. I wasn't using a Shiny modal, but I was using Shiny modules. Thanks to @gadenbuie's thorough evaluation of the issue, especially the sentence "you cannot call updateSelectInput() to update the element in the modal, until you know that the modal has been shown", I realised that - in my case at least - I might have been calling the server functions of my modules prior to calling their UI functions, leading to the default values of the UI components overriding any values set to the components by the server functions. I checked, and it was indeed the case. What works for me now is this:
1. Call the UI function of the module, but don't render the result yet.
2. Call the server function of the module.
3. Render the result of the UI function of the module.

Just thought I'll post this here to add it to the public knowledge base :)

Thanks @gadenbuie!
